govcd/metadata_v2.go

```diff
@@ -848,23 +868,149 @@ func mergeMetadataAndWait(client *Client, requestUri string, metadata map[string
-func deleteMetadataAndWait(client *Client, requestUri string, key string, isSystem bool) error {
-	task, err := deleteMetadata(client, requestUri, key, isSystem)
+func deleteMetadataAndWait(client *Client, requestUri, name, key string, isSystem bool) error {
+	task, err := deleteMetadata(client, requestUri, name, key, isSystem)
 	if err != nil {
 		return err
 	}
 
 	return task.WaitTaskCompletion()
 }
+
+// IgnoredMetadata is a structure that defines the metadata entries that should be ignored by the VCD Client.
+// The filtering works in such a way that all the non-nil pointers in an instance of this struct are evaluated with
+// a logical AND.
+// For example, ignoredMetadata.ObjectType = "org", ignoredMetadata.ObjectName = "foo" will ignore all metadata from
+// Organizations whose name is "foo", with any key and any value.
+// Note: This struct is only used by metadata_v2.go methods.
+// Note 2: Filtering by ObjectName is not possible in the "ByHref" methods from VCDClient.
+type IgnoredMetadata struct {
+	ObjectType *string        // Type of the object that has the metadata as defined in the API documentation https://developer.vmware.com/apis/1601/vmware-cloud-director, for example "catalog", "disk", "org"...
+	ObjectName *string        // Name of the object
+	KeyRegex   *regexp.Regexp // A regular expression to filter out metadata keys
+	ValueRegex *regexp.Regexp // A regular expression to filter out metadata values
+}
+
+// filterMetadata filters all metadata entries, given a slice of metadata that needs to be ignored. It doesn't
+// alter the input metadata, but returns a copy of the filtered metadata.
+func filterMetadata(allMetadata *types.Metadata, href, name string, metadataToIgnore []IgnoredMetadata) (*types.Metadata, error) {
+	if len(metadataToIgnore) == 0 {
+		return allMetadata, nil
+	}
+
+	result := &types.Metadata{
+		XMLName:       allMetadata.XMLName,
+		Xmlns:         allMetadata.Xmlns,
+		HREF:          allMetadata.HREF,
+		Type:          allMetadata.Type,
+		Xsi:           allMetadata.Xsi,
+		Link:          allMetadata.Link,
+		MetadataEntry: nil,
+	}
+
+	var filteredMetadata []*types.MetadataEntry
+	for _, originalEntry := range allMetadata.MetadataEntry {
+		_, err := filterSingleMetadataEntry(originalEntry.Key, href, name, &types.MetadataValue{Domain: originalEntry.Domain, TypedValue: originalEntry.TypedValue}, metadataToIgnore)
+		if err == nil || !strings.Contains(err.Error(), "ignored") {
+			filteredMetadata = append(filteredMetadata, originalEntry)
+		}
+	}
+	result.MetadataEntry = filteredMetadata
+	return result, nil
+}
+
+// filterSingleMetadataEntry filters a single metadata entry given a slice of metadata that needs to be ignored. It doesn't
+// alter the input metadata, but returns a copy of the filtered metadata.
+func filterSingleMetadataEntry(key, href, name string, metadataEntry *types.MetadataValue, metadataToIgnore []IgnoredMetadata) (*types.MetadataValue, error) {
```

Comment:
I found the `name` to be ambiguous here (and to be fair in other places) while reading through the code until I finally got to the comparison itself in https://github.com/vmware/go-vcloud-director/pull/581/files#diff-3e1319d45ed96bc5f0fb18bd53192105cb8238939c6c50c3763aad3b82d319cbR956

```suggestion
func filterSingleMetadataEntry(key, href, objectName string, metadataEntry *types.MetadataValue, metadataToIgnore []IgnoredMetadata) (*types.MetadataValue, error) {
```

Reply:
True, changed